### DISCUSSION TITLE
is there a way to retrieve all AST Node ? (i.e. browse eAllContents() ?)

### DISCUSSION THREAD

**Discussion by jdeantoni** [eclipse-langium/langium#944] [Q&A] [answered]
Hi,

I was wondering if there is an equivalent to the [EMF eAllContents()](https://download.eclipse.org/modeling/emf/emf/javadoc/2.5.0/org/eclipse/emf/ecore/EObject.html#eAllContents()) or to the QVTo allSubObjects() that would help to browse through all contained nodes from a specific node of the graph ?

**Comment by msujew** [comment, marked as answer]
Hey @jdeantoni,

we have a few utility functions for this case. Namely [`streamAst`](https://github.com/langium/langium/blob/0808e7e21bcf8f648e6eab5c54ed5bc2f5024776/packages/langium/src/utils/ast-util.ts#L159), [`streamContents`](https://github.com/langium/langium/blob/0808e7e21bcf8f648e6eab5c54ed5bc2f5024776/packages/langium/src/utils/ast-util.ts#L107) and [`streamAllContents`](https://github.com/langium/langium/blob/0808e7e21bcf8f648e6eab5c54ed5bc2f5024776/packages/langium/src/utils/ast-util.ts#L148). Refer to their documentation to see what each of these functions does in particular.

**Reply by jdeantoni** [reply]
Thanks  @msujew ! sorry I did not find them :)
